### PR TITLE
Add support for excluding coins in `create_signed_transaction` wallet rpc

### PR DESCRIPTION
**Pull request by felixbrucker**
Currently it is not possible to exclude specific coins in the `create_signed_transaction` wallet rpc api, only to select specific coins (allow vs deny list).

This PR adds a new option `exclude_coins` to the `create_signed_transaction` wallet rpc api, which gets passed down into the `select_coins()` method, which already supports exclusion of coins.

This PR is missing tests because i don't understand the current testing methods and their dependencies, it seems all tests use data of previous tests and are not contained in single test methods with each their own setup and teardown. If possible i'd want to add tests ofc.

### CODE REVIEW

File: chia/wallet/wallet.py [hunk 1]
```diff
@@ -332,7 +333,10 @@ async def _generate_unsigned_transaction(
                 raise ValueError(f"Can't send more than {max_send} in a single transaction")
 
         if coins is None:
-            coins = await self.select_coins(total_amount)
+            exclude_coins_list: Optional[List[Coin]] = None
```
Comment (aqk) [inline chia/wallet/wallet.py:336]: We'll need a test to exercise these cases.

**Review by aqk — DISMISSED** [review]

**Comment by aqk** [comment]
This is looking great. Two things I'd like to see before we land it:

* Update docs about intended behaviour. Right now, this patch only uses the `exclude_coins` if `coins` is `None` (but not `[]`)

* Add tests to exercise at least these cases: exclude_coins in [None, [], some] x coins in [None, [], some]


@mariano54

File: chia/wallet/wallet.py [hunk 2]
```diff
@@ -332,7 +333,10 @@ async def _generate_unsigned_transaction(
...
+            if exclude_coins is not None:
```
[the remark this answers is not in the document]
Comment (mariano54) [reply chia/wallet/wallet.py:337]: The first line of the function should be something like:
```
if exclude_coins is: None: exclude_coins = []
if coins is None: coins = []
```
That way, there is no ambiguity about things being None or empty. 

I think it's reasonable to be able to use both, but this should be tested explicitly, and if there is a conflict, it should throw an error. Ignoring what the user passed in is a bad idea and will lead to bugs and complaints.

**Review by mariano54 — DISMISSED** [review]

File: chia/wallet/wallet.py (hunk 1 above)
Comment (felixbrucker) [reply chia/wallet/wallet.py:336]: Right. I'd love to add some tests, but as i stated above im a bit lost on where tests start and end, currently they seem to be a little messy. Any pointers are appreciated 👍

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:337]: My reason for only using `exclude_coins` when selecting coins is that when a user wants to select specific coins via `coins`, he can already filter those client side, as the inputs for both are the same: `Coin`s. That being said, we can ofc also do a server side filter of `coins` by `exclude_coins` if both are set.

When also removing the optional part of both options we will have to throw early in case the server side filter leaves no coins in the array, because otherwise we'd fallback to auto selecting, because now we only check for empty.

> and if there is a conflict, it should throw an error

what do you mean by that, i mean how would they conflict?

**Comment by felixbrucker** [comment]
Still waiting for answers..

**Comment by emlowe** [comment]
I would suggest adding a test in `test_coin_selection.py` that exercises the exclude option of `select_coins()` (this should really already be there, but it's not)
You could also add a test to `test_wallet_rpc.py` if you look at `test_create_signed_transaction` you can probably replicate much of that test and modify as needed

**Comment by felixbrucker** [comment]
@emlowe Thanks, i added some tests in these files 👍

**Comment by wallentx** [comment]
Hey @felixbrucker, sorry about this. We've been pretty busy the past few weeks, and we haven't done a great job of budgeting the needed time to tending to community PRs. There's definitely been some interest in this particular PR, and we would love to work with you to get this in if you are willing to reopen it.

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (emlowe) [reply chia/wallet/wallet.py:317]: I believe the suggestion is talking about what happens when a non-empty set of `coins` and a non-empty set of `exclude_coins` are both passed in. This PR ignores in this case the `exclude_coins` which I think is what Mariano is talking about. But if you don't ignore it, then what happens if the same coin is in both `coins` and `exclude_coins` - to which the suggestion was to raise an exception. Somewhat edge case, also very easy to check in python since they are both Sets `bool(coins & excluded_coins)`

**Comment by jack60612** [comment]
similar changes in this pr too https://github.com/Chia-Network/chia-blockchain/pull/12360, if you want a point of reference

File: chia/rpc/wallet_rpc_api.py [hunk 3]
```diff
@@ -1696,6 +1696,10 @@ async def create_signed_transaction(self, request, hold_lock=True) -> EndpointRe
         if "coins" in request and len(request["coins"]) > 0:
             coins = set([Coin.from_json_dict(coin_json) for coin_json in request["coins"]])
 
+        exclude_coins = None
+        if "exclude_coins" in request and len(request["exclude_coins"]) > 0:
+            exclude_coins = set([Coin.from_json_dict(coin_json) for coin_json in request["exclude_coins"]])
```
Comment (jack60612) [inline chia/rpc/wallet_rpc_api.py:1701]: exclude_coins is a list, not a set.

File: chia/rpc/wallet_rpc_client.py [hunk 4]
```diff
@@ -246,6 +247,10 @@ async def create_signed_transaction(
             coins_json = [c.to_json_dict() for c in coins]
             request["coins"] = coins_json
 
+        if exclude_coins is not None and len(exclude_coins) > 0:
+            exclude_coins_json = [exclude_coin.to_json_dict() for exclude_coin in exclude_coins]
+            request["exclude_coins"] = exclude_coins_json
```
Comment (jack60612) [inline chia/rpc/wallet_rpc_client.py:252]: why are we doing this?

File: chia/wallet/wallet.py [hunk 5]
```diff
@@ -291,6 +291,7 @@ async def _generate_unsigned_transaction(
         memos: Optional[List[bytes]] = None,
         negative_change_allowed: bool = False,
         min_coin_amount: Optional[uint128] = None,
+        exclude_coins: Optional[Set[Coin]] = None,
```
Comment (jack60612) [inline chia/wallet/wallet.py:294]: why is this not just `Optional[List[Coin]]`? solves a lot of problems

File: chia/wallet/wallet.py [hunk 6]
```diff
@@ -418,6 +424,7 @@ async def generate_signed_transaction(
...
```
Comment (jack60612) [inline chia/wallet/wallet.py:427]: List please

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/wallet.py:317]: I say after making it a list, we do if `coins is not None and excluded_coins is not None:` make a copy of coins & a new list and loop through the list only including the non excluded coins. ex:
`old_coins = coins`
`coins: List[Coin] = [] for coin in old_coins:`
       `    if coin is not in excluded_coins:`
        `coins.append(coin)`

**Review by jack60612 — CHANGES_REQUESTED** [review]
Made a couple of comments to simplify the logic and make it more efficient. I put another pr that did something similar below if you need any code to look at.

File: chia/wallet/wallet.py (hunk 5 above)
Comment (felixbrucker) [reply chia/wallet/wallet.py:294]: This is a set to be consistent with `coins` from above

Comment (jack60612) [reply chia/wallet/wallet.py:294]: Why dont we just make everything a list?

Comment (felixbrucker) [reply chia/wallet/wallet.py:294]: Good question, i found it a bit weird to constantly switch between set and list (both have their up and downsides).
I dont really want to make such "huge" changes, those are probably better suited for a refactoring pr unrelated to this issue that uses set or list consistently across the codebase for coin sets/lists

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:317]: @emlowe Is the expected behavior of the method that when passing both in we do a filter of the `coins` using `exclude_coins` first thing? in that case we wouldnt need to throw as well. If no, then what is the expected behavior in this scenario?

File: chia/wallet/wallet.py (hunk 5 above)
Comment (jack60612) [reply chia/wallet/wallet.py:294]: > Good question, i found it a bit weird to constantly switch between set and list (both have their up and downsides). I dont really want to make such "huge" changes, those are probably better suited for a refactoring pr unrelated to this issue that uses set or list consistently across the codebase for coin sets/lists

I meant just for exclude_coins, i think the normal coins should stay a set for efficiency reasons

Comment (felixbrucker) [reply chia/wallet/wallet.py:294]: Having different types feels weird to me

Comment (jack60612) [reply chia/wallet/wallet.py:294]: Unless you think exclude_coins will end up quite large, which then a set makes sense.

Comment (felixbrucker) [reply chia/wallet/wallet.py:294]: i mean `coins` and `exclude_coins` are basically the same thing

Comment (jack60612) [reply chia/wallet/wallet.py:294]: > Having different types feels weird to me

in my mind longer sets are better then longer lists & sets are a bit faster to loop through.

Comment (jack60612) [reply chia/wallet/wallet.py:294]: > i mean `coins` and `exclude_coins` are basically the same thing

true, but generally exclude_coins will be smaller than coins.
Please don't switch between sets and lists in this pr though, just use lists and this can be a discussion for later

Comment (felixbrucker) [reply chia/wallet/wallet.py:294]: I kept the api consistent, `exclude_coins` should be the same type as `coins`, anything else is unexpected.

Comment (jack60612) [reply chia/wallet/wallet.py:294]: > I kept the api consistent, `exclude_coins` should be the same type as `coins`, anything else is unexpected.

fair, ill refactor it later if need be.

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:317]: The following piece of code could be added below the coin selection condition:
```
elif exclude_coins is not None:
    if bool(coins & exclude_coins):
        raise ValueError(f"Can't exclude coins specifically included")
    coins = coins - exclude_coins
```
Depending on what the expected behavior of the method is when given both coins and exclude_coins we can also remove the throw before the filter

File: chia/rpc/wallet_rpc_client.py (hunk 4 above)
Comment (felixbrucker) [reply chia/rpc/wallet_rpc_client.py:252]: I used the same approach as done just a few lines above for `coins`

File: chia/rpc/wallet_rpc_api.py (hunk 3 above)
Comment (felixbrucker) [reply chia/rpc/wallet_rpc_api.py:1701]: `exclude_coins` gets passed into `generate_signed_transaction()` which accepts a set, or do you mean something else?

Comment (jack60612) [reply chia/rpc/wallet_rpc_api.py:1701]: sorry i mixed up the api and the client when reviewing

File: chia/wallet/wallet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/wallet.py:317]: I would remove the throw but put a warning saying that this is not recommended.

[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:317]: > I would remove the throw but put a warning saying that this is not recommended.

But if its the expected behavior why warn?

[the remark this answers is not in the document]
Comment (jack60612) [reply chia/wallet/wallet.py:317]: > > I would remove the throw but put a warning saying that this is not recommended.
> 
> But if its the expected behavior why warn?

fair point, i think it should error out then.

[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:317]: Yes, i agree. So the expected behavior is that when passing both and they intersect we throw. One might even throw directly, even if they dont intersect, because there is no valid case where they are ever used. What do you think?

[the remark this answers is not in the document]
Comment (felixbrucker) [reply chia/wallet/wallet.py:317]: Updated the PR to reflect that :+1:

**Review by jack60612 — DISMISSED** [review]
LGTM on test pass
coin_selection tests seem fine to me

**Comment by felixbrucker** [comment]
General note on tests: i noticed there are a few flaky (unrelated) tests, however i cant restart failed tests on gha. All recent master commits have failing tests as well.

**Comment by jack60612** [comment]
> General note on tests: i noticed there are a few flaky (unrelated) tests, however i cant restart failed tests on gha. All recent master commits have a few failing tests as well.

no worries, flake is flake.

**Comment by felixbrucker** [comment]
@jack60612 3 unrelated tests are failing

**Comment by jack60612** [comment]
> This pull request has conflicts, please resolve those before we can evaluate the pull request.

sorry :)

**Comment by felixbrucker** [comment]
@jack60612 i noticed we now have inconsistent naming: `exclude` is used throughout the codebase, i added `exclude_coins` to be consistent to that. Now with the latest PRs `excluded_coins` was added as well, so now we have an inconsistent mix.
For users and devs it would probably be best if naming were consistent

**Comment by felixbrucker** [comment]
Ah, ofc main is failing with 70+ test fails :sweat_smile: 

someone please ping me when main is safe to rebase onto again

**Comment by jack60612** [comment]
> Ah, ofc main is failing with 70+ test fails :sweat_smile: 
> 
> 
> 
> someone please ping me when main is safe to rebase onto again

Try it now

**Review by jack60612 — APPROVED** [review]
lgtm

**Review by emlowe — APPROVED** [review]

**Review by wjblanke — APPROVED** [review]
aok
